### PR TITLE
Lock traffic notifications

### PR DESCRIPTION
**Pull request by aerosol**
### Changes

~~:warning: This PR contains ~~two migrations~~ one migration that will be extracted to a separate PR once the review is done.~~

`accept_traffic_until` (now on the `User` record) is monitored on a daily basis. Whenever that date approaches by either 1 day or 7 days, we'll send out notifications to owners, assuming their sites have any traffic still - with the predicate for that check being any event recorded "since yesterday".

`accept_traffic_until` is updated whenever a billing event arrives (subscription either updated or changed) or effectively, or whenever a new trial is started.

This PR won't go live before backfilling the existing data first, hence keeping it as a draft for now.

`accept_traffic_until` is exposed in the CRM too. Whenever `trial_expiry_date` is changed, it is bumped along the way accordingly.

### Tests
- [x] Automated tests have been added
- [ ] This PR does not require tests

### Changelog
- [ ] Entry has been added to changelog
- [x] This PR does not make a user-facing change

### Documentation
- [ ] [Docs](https://github.com/plausible/docs) have been updated
- [x] This change does not need a documentation update

### Dark mode
- [ ] The UI has been tested both in dark and light mode
- [x] This PR does not change the UI

### CODE REVIEW

File: lib/workers/accept_traffic_until_notification.ex [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    )
+  end
+
+  defp store_sent(notification, today) do
+    Repo.insert_all("sent_accept_traffic_until_notifications", [
+      %{
+        user_id: notification.id,
+        sent_at: today
+      }
+    ])
```
Comment (zoldar) [inline lib/workers/accept_traffic_until_notification.ex:91]: Perhaps turning that into an upsert would be safer? Failing insert may result in sending the email more than once.

```suggestion
    Repo.insert_all("sent_accept_traffic_until_notifications", [
      %{
        user_id: notification.id,
        sent_at: today
      }
    ], on_conflict: :nothing, conflict_target: [:user_id, :sent_at])
```

File: lib/workers/accept_traffic_until_notification.ex [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+  @impl Oban.Worker
+  def perform(_job, today \\ Date.utc_today()) do
+    tomorrow = today |> Date.add(+1)
+    next_week = today |> Date.add(+7)
+
+    # send at most one notification per user, per day
+    sent_today =
+      from s in "sent_accept_traffic_until_notifications",
+        where: s.sent_at == ^today
```
[the remark this answers is not in the document]
Comment (aerosol) [reply lib/workers/accept_traffic_until_notification.ex:27]: This is not per site, but per user

[the remark this answers is not in the document]
Comment (aerosol) [reply lib/workers/accept_traffic_until_notification.ex]: Maybe 9859995ec ?

File: priv/repo/migrations/20231219083050_track_accept_traffic_until_notifcations.exs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+defmodule Plausible.Repo.Migrations.TrackAcceptTrafficUntilNotifcations do
+  use Ecto.Migration
+
+  def change do
+    create table(:sent_accept_traffic_until_notifications) do
+      add :user_id, references(:users, on_delete: :delete_all), null: false
+      add :sent_at, :date, null: false
```
[the remark this answers is not in the document]
Comment (aerosol) [reply priv/repo/migrations/20231219083050_track_accept_traffic_until_notifcations.exs]: 340bd63e9

File: lib/workers/accept_traffic_until_notification.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (zoldar) [reply lib/workers/accept_traffic_until_notification.ex]: Yup, looks good! Sorry for mixing it up 🤦

File: lib/workers/accept_traffic_until_notification.ex (hunk 1 above)
Comment (aerosol) [reply lib/workers/accept_traffic_until_notification.ex]: 8859eafb6

File: lib/plausible/billing/billing.ex [hunk 4]
```diff
@@ -271,6 +271,7 @@ defmodule Plausible.Billing do
 
     user
     |> maybe_remove_grace_period()
+    |> tap(&Plausible.Sites.update_accept_traffic_until/1)
```
Comment (zoldar) [inline lib/plausible/billing/billing.ex:274]: Looking at - https://github.com/plausible/analytics/pull/3641/files#diff-6a341128111d11eddb083bfa680c64e493749d12777e2184d186a84d04e83167R212-R220 I'm wondering if `last_bill_date` can ever end up empty/null? 🤔

Comment (aerosol) [reply lib/plausible/billing/billing.ex:274]: it's `next_bill_date` and it's never null unless the plan is free

File: lib/plausible/sites.ex [hunk 5]
```diff
@@ -13,6 +14,7 @@ defmodule Plausible.Sites do
 
   require Plausible.Site.UserPreference
 
+  @accept_traffic_until_free ~D[2035-01-01]
```
Comment (ukutaht) [inline lib/plausible/sites.ex:17]: 11 years is a long time but it's realistic that someone will actually hit this limit 😀 Something like 100 years in the future would be safer

Comment (aerosol) [reply lib/plausible/sites.ex]: Love it lol. Done in https://github.com/plausible/analytics/pull/3641/commits/87c4481066b5030d5dda03ec3ba68e853742915d

**Comment by aerosol** [comment]
@zoldar @ukutaht pushed the last round of updates including the CRM changes; updated the PR description. This is ready for review as far as I'm concerned.

File: lib/plausible/auth/user.ex [hunk 6]
```diff
@@ -107,12 +111,24 @@ defmodule Plausible.Auth.User do
...
     |> validate_required([:email, :name, :email_verified])
+    |> maybe_bump_accept_traffic_until()
     |> unique_constraint(:email)
   end
 
+  defp maybe_bump_accept_traffic_until(changeset) do
+    expiry_change = get_change(changeset, :trial_expiry_date)
+
+    if expiry_change do
+      put_change(changeset, :accept_traffic_until, Date.add(expiry_change, 14))
```
Comment (zoldar) [inline lib/plausible/auth/user.ex:126]:
```suggestion
      put_change(changeset, :accept_traffic_until, Date.add(expiry_change, @trial_accept_traffic_until_offset_days))
```
?

File: lib/plausible/site/admin.ex [hunk 7]
```diff
@@ -117,11 +128,16 @@ defmodule Plausible.SiteAdmin do
     Timex.format!(date, "{Mshort} {D}, {YYYY}")
   end
 
-  defp get_owner_email(site) do
+  defp get_owner(site) do
     owner = Enum.find(site.memberships, fn m -> m.role == :owner end)
```
Comment (zoldar) [inline lib/plausible/site/admin.ex:132]: If we add `:owner` to preloads at https://github.com/plausible/analytics/pull/3641/files#diff-ffff0c95718b914d284cc320ae6be63fe3da9bddcab1d81abaf0f13f7b8aece5R17, we could change that to

```suggestion
    owner = site.owner
```

File: lib/plausible/site/admin.ex [hunk 8]
```diff
@@ -39,15 +39,26 @@ defmodule Plausible.SiteAdmin do
...
-            n -> "⏱ #{n}/#{site.ingest_rate_limit_scale_seconds}s (per server)"
-          end
+          rate_limiting_status =
+            case site.ingest_rate_limit_threshold do
+              nil -> ""
+              0 -> "🛑 BLOCKED"
+              n -> "⏱ #{n}/#{site.ingest_rate_limit_scale_seconds}s (per server)"
+            end
+
+          owner = Enum.find(site.memberships, fn m -> m.role == :owner end)
```
Comment (zoldar) [inline lib/plausible/site/admin.ex:53]: Like mentioned elsewhere, with :owner preload this becomes

```suggestion
          owner = site.owner
```

File: lib/plausible/users.ex [hunk 9]
```diff
@@ -2,13 +2,51 @@ defmodule Plausible.Users do
   @moduledoc """
   User context
   """
+  use Plausible
+  @accept_traffic_until_free ~D[2135-01-01]
```
Comment (zoldar) [inline lib/plausible/users.ex:6]: image

File: lib/plausible_web/templates/email/trial_over_email.html.eex [hunk 10]
```diff
@@ -3,4 +3,4 @@ Your free Plausible trial has now expired. Upgrade your account to continue rece
 <%= link("Upgrade now", to: PlausibleWeb.Router.Helpers.billing_url(PlausibleWeb.Endpoint, :choose_plan)) %>
 <br /><br />
 
-We will keep recording stats for another month to give you time to upgrade.
+We will keep recording stats for 14 days to give you time to upgrade.
```
Comment (zoldar) [inline lib/plausible_web/templates/email/trial_over_email.html.eex:6]: Maybe parametrize these 14 days using the exposed attribute from User?

File: test/support/test_utils.ex [hunk 11]
```diff
@@ -213,6 +213,10 @@ defmodule Plausible.TestUtils do
     NaiveDateTime.truncate(naive, :second)
   end
 
+  def to_naive_truncate(%Date{} = date) do
```
Comment (zoldar) [inline test/support/test_utils.ex:216]: 👍

**Review by zoldar — APPROVED** [review]
✨

File: lib/plausible/site/admin.ex (hunk 7 above)
Comment (aerosol) [reply lib/plausible/site/admin.ex]: https://github.com/plausible/analytics/pull/3641/commits/ca390f08842c94276522831ede13f9bf68e818c3

I checked and there's no owner-less sites anymore

File: lib/plausible_web/templates/email/trial_over_email.html.eex (hunk 10 above)
Comment (aerosol) [reply lib/plausible_web/templates/email/trial_over_email.html.eex]: 4563c2466

**Pull request merged**